pkg/provisioner/ironic/ironic.go

```diff
@@ -1734,46 +1734,66 @@ func (p *ironicProvisioner) PowerOn() (result provisioner.Result, err error) {
 			p.log.Info("waiting for power status to change")
 			return operationContinuing(powerRequeueDelay)
 		}
-		result, err = p.changePower(ironicNode, nodes.PowerOn)
-		switch err.(type) {
-		case nil:
-		case HostLockedError:
-		default:
-			return transientError(errors.Wrap(err, "failed to power on host"))
+		if ironicNode.LastError != "" && !force {
+			p.log.Info("PowerOn operation failed", "msg", ironicNode.LastError)
+			return operationFailed(fmt.Sprintf("PowerOn operation failed: %s",
+				ironicNode.LastError))
+		}
+		if result, err = p.changePower(ironicNode, nodes.PowerOn); err == nil {
+			p.publisher("PowerOn", "Host powered on")
+			return result, nil
+		} else {
+			switch err.(type) {
+			case HostLockedError:
+				return retryAfterDelay(powerRequeueDelay)
+			default:
+				return transientError(errors.Wrap(err, "failed to PowerOn node"))
+			}
 		}
-		p.publisher("PowerOn", "Host powered on")
 	}
-
 	return result, nil
 }
 
 // PowerOff ensures the server is powered off independently of any image
 // provisioning operation.
-func (p *ironicProvisioner) PowerOff(rebootMode metal3v1alpha1.RebootMode) (result provisioner.Result, err error) {
+func (p *ironicProvisioner) PowerOff(rebootMode metal3v1alpha1.RebootMode, forceHardPowerOff bool) (result provisioner.Result, err error) {
 	p.log.Info(fmt.Sprintf("ensuring host is powered off (mode: %s)", rebootMode))
 
-	if rebootMode == metal3v1alpha1.RebootModeHard {
-		result, err = p.hardPowerOff()
+	if rebootMode != metal3v1alpha1.RebootModeHard && !forceHardPowerOff {
+		if result, err = p.softPowerOff(); err == nil {
+			return result, nil
+		} else {
+			switch err.(type) {
+			// In case of soft power off is unsupported or has failed,
+			// we activate hard power off.
+			case SoftPowerOffUnsupportedError, SoftPowerOffFailed:
+				p.log.Info("soft power off is unsupported or failed. Proceeding to hard power off.")
+				return operationFailed(result.ErrorMessage)
+			case HostLockedError:
```

Comment:
Maybe this should be a separate commit, but I actually think now we could get rid of all these error types and just returned the `result`/`error` combination we want at the source and ditch this whole switch statement to just `return p.softPowerOff()` (and likewise for `hardPowerOff()`.

Reply:
Fixed as part of https://github.com/metal3-io/baremetal-operator/pull/841/commits/3d834e33e9b764f21298cd6a3661164133529f8d.